pkg/cmd/pulumi/config.go

```diff
@@ -80,10 +80,165 @@ func newConfigCmd() *cobra.Command {
+		Args: cmdutil.MaximumNArgs(1),
+		Run: cmdutil.RunFunc(func(cmd *cobra.Command, args []string) error {
+			opts := display.Options{
+				Color: cmdutil.GetGlobalColorization(),
+			}
+
+			// Get current stack and ensure that it is a different stack to the destination stack
+			currentStack, err := requireStack(*stack, false, opts, true /*setCurrent*/)
+			if err != nil {
+				return err
+			}
+			if currentStack.Ref().Name().String() == destinationStackName {
+				return errors.New("config must be copied to a different destination stack than currently using")
+			}
+			currentProjectStack, err := loadProjectStack(currentStack)
+			if err != nil {
+				return err
+			}
+
+			// Get the destination stack
+			destinationStack, err := requireStack(destinationStackName, false, opts, false /*setCurrent*/)
+			if err != nil {
+				return err
+			}
+			destinationProjectStack, err := loadProjectStack(destinationStack)
+			if err != nil {
+				return err
+			}
+
+			// Do we need to copy a single value or the entire map
+			if len(args) > 0 {
+				// A single key was specified so we only need to copy that specific value
+				return copySingleConfigKey(args[0], path, currentStack, currentProjectStack, destinationStack,
+					destinationProjectStack)
+			}
+
+			return copyEntireConfigMap(currentStack, currentProjectStack, destinationStack, destinationProjectStack)
+		}),
+	}
+
+	cpCommand.PersistentFlags().BoolVar(
+		&path, "path", false,
+		"The key contains a path to a property in a map or list to set")
+	cpCommand.PersistentFlags().StringVarP(
+		&destinationStackName, "new-stack", "", "",
+		"The name of the new stack to copy the config to")
+
+	return cpCommand
+}
+
+func copySingleConfigKey(configKey string, path bool, currentStack backend.Stack,
+	currentProjectStack *workspace.ProjectStack, destinationStack backend.Stack,
+	destinationProjectStack *workspace.ProjectStack) error {
+	var decrypter config.Decrypter
+	key, err := parseConfigKey(configKey)
+	if err != nil {
+		return errors.Wrap(err, "invalid configuration key")
+	}
+
+	v, ok, err := currentProjectStack.Config.Get(key, path)
+	if err != nil {
+		return err
+	}
+	if ok {
+		if v.Secure() {
+			var err error
+			if decrypter, err = getStackDecrypter(currentStack); err != nil {
+				return errors.Wrap(err, "could not create a decrypter")
+			}
+		} else {
+			decrypter = config.NewPanicCrypter()
+		}
+
+		raw, err := v.Value(decrypter)
+		if err != nil {
+			return errors.Wrap(err, "could not decrypt configuration value")
+		}
+
+		// now we start the work to copy it - if it was encrypted before then we need to do the same again
+		if v.Secure() {
+			// it was encrypted before, now we need to re-encrypt
+			c, cerr := getStackEncrypter(destinationStack)
+			if cerr != nil {
+				return cerr
+			}
+			enc, eerr := c.EncryptValue(raw)
+			if eerr != nil {
+				return eerr
+			}
+			v = config.NewSecureValue(enc)
+		} else {
+			v = config.NewValue(raw)
+		}
```

Comment:
This doesn't appear to be handling objects correctly.

For example:

```
$ pulumi stack init dev
$ pulumi stack init dev2
$ pulumi stack select dev
$ pulumi config set aws:region us-west-2
$ pulumi config set --path data.active true
$ pulumi config set --path data.nums[0] 1
$ pulumi config set --path data.nums[1] 2
$ pulumi config set --path data.nums[2] 3
$ pulumi config set --path data.pass --secret foo
$ cat Pulumi.dev.yaml
config:
  aws:region: us-west-2
  stackcopytestage:data:
    active: true
    nums:
    - 1
    - 2
    - 3
    pass:
      secure: AAABANIHq5zNQDqx8G/SHeR1aehVjWKUsv9pIMrXG/z9C00=
$ pulumi config cp data --new-stack dev2
$ cat Pulumi.dev2.yaml
config:
  stackcopytestage:data:
    secure: AAABAFFV4QQd9AxeUeC4wSxZ6ST6zYIR0wIr+pmLBQg7Uxi0NYrwV/SZ/rnClAcMOvhqAof0DKbrtHfP6kACT79gPUF7VHhySU8c
```

I would expect `Pulumi.dev2.yaml` to look like `Pulumi.dev.yaml`, just with a different encrypted value for `pass`:

```
config:
  aws:region: us-west-2
  stackcopytestage:data:
    active: true
    nums:
    - 1
    - 2
    - 3
    pass:
      secure: <some other encrypted string>
```

Right now, it's encrypting the whole config value, not the individual `pass` property inside the object, because it's not taking `if c.Object() {` into account.

I wonder if it'd be easier to add a low-level `Copy(decrypter Decrypter, encrypter Encrypter)` to `Value` in `sdk/go/common/resource/config/value.go` that handles this, and then use `Value.Copy` both here and inside `Map.Copy`.

Reply:
Fixed and working as expected now!